scripts/lint-mdc.ts

```diff
@@ -496,28 +757,234 @@ function checkOneClassMemberIsInReadme(
+        collectPubliclyExportedClassNames(nodePath, publiclyExportedClassNames);
+      },
+      InterfaceDeclaration(nodePath: NodePath<babelTypes.InterfaceDeclaration>) {
+        checkOneAdapterName(nodePath, inputFilePath);
+      },
+      TSInterfaceDeclaration(nodePath: NodePath<babelTypes.TSInterfaceDeclaration>) {
+        checkOneAdapterName(nodePath, inputFilePath);
+      },
+    });
+
+    if (!publiclyExportedClassNames.some((className) => className.endsWith('Adapter'))) {
+      logLinterViolation(
+          inputFilePath,
+          ast.loc,
+          `*adapter.ts files must export at least one public interface ending in 'Adapter'.`,
+      );
+      return;
+    }
+  }
+
+  if (isFoundationFile) {
+    if (!publiclyExportedClassNames.some((className) => className.endsWith('Foundation'))) {
+      logLinterViolation(
+          inputFilePath,
+          ast.loc,
+          `All *foundation.ts files must export at least one public class ending in 'Foundation'.`,
+      );
+      return;
+    }
+  }
+}
+
+function checkOneTopLevelAdapterNode(nodePath: NodePath, inputFilePath: string) {
+  const node = nodePath.node;
+  const {name, loc, kind} = getIdentifierInfo(node);
+
+  const isFileLevelNode =
+      nodePath.isFile() ||
+      nodePath.isProgram() ||
+      nodePath.isExportDeclaration() ||
+      isImportDeclaration(nodePath);
+  const isInterface = babelTypes.isInterfaceDeclaration(node) || babelTypes.isTSInterfaceDeclaration(node);
+  if (isFileLevelNode || isInterface) {
+    return;
+  }
+
+  logLinterViolation(
+      inputFilePath,
+      loc,
+      [
+        `Invalid top-level ${kind} (${node.type})${name ? `: '${name}'` : ''}.`,
+        `Only interfaces are allowed in *adapter.ts files.`,
+      ],
+  );
+}
+
+function checkOneAdapterName(nodePath: NodePath<ClassLikeNode>, inputFilePath: string) {
+  const node = nodePath.node;
+  const {name, loc, kind} = getIdentifierInfo(node);
+  const id = nodePath.node.id;
+
+  if (id && !id.name.endsWith('Adapter')) {
+    logLinterViolation(
+        inputFilePath,
+        loc,
+        [
+          `Invalid top-level ${kind} (${node.type})${name ? `: '${name}'` : ''}.`,
+          `All type names in *adapter.ts files must end with 'Adapter'.`,
+          `This keeps the files lean and predictable.`,
+        ],
+    );
+    return;
+  }
+
+  if (!isInlineExportedSymbol(nodePath)) {
+    logLinterViolation(
+        inputFilePath,
+        loc,
+        [
+          `Invalid top-level ${kind} (${node.type})${name ? `: '${name}'` : ''}.`,
+          `All types in *adapter.ts files must be exported.`,
+        ],
+    );
+    return;
+  }
+}
+
+function collectPubliclyExportedClassNames(nodePath: NodePath<ClassLikeNode>, classNames: string[]) {
+  const id = nodePath.node.id;
+  if (id && isInlineExportedSymbol(nodePath)) {
+    classNames.push(id.name);
+  }
+}
+
+function getIdentifierInfo(node: babelTypes.Node): NamedIdentifier {
```

Comment:
This seems like a lot of code to normalize across different things in the AST... I guess there's no built in convenience for anything like this?